File: src/24.两两交换链表中的节点.py
#
# @lc app=leetcode.cn id=24 lang=python3
#
# [24] 两两交换链表中的节点
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def swapPairs(self, head: ListNode) -> ListNode:
        new_head = ListNode(0)
        new_head.next = head
        p1, p2 = new_head, new_head.next
        while p1 and p2:
            t = p2.next
            if t:
                m = t.next
                p1.next = t
                t.next = p2
                p2.next = m
            p1, p2 = p2, p2.next
        return new_head.next
# @lc code=end

File: src/24.两两交换链表中的节点.go
package main

/*
 * @lc app=leetcode.cn id=24 lang=golang
 *
 * [24] 两两交换链表中的节点
 */
// linked-list

import "fmt"

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

func swapPairs(head *ListNode) *ListNode {
	ans := &ListNode{0, head}
	for p1, p2 := ans, ans.Next; p1 != nil && p2 != nil && p2.Next != nil ; p1, p2 = p2, p2.Next {
		t := p2.Next
		p2.Next = t.Next
		t.Next = p2
		p1.Next = t
	}
	return ans.Next
}

// @lc code=end
func main() {
	fmt.Println(swapPairs(&ListNode{1, &ListNode{2, &ListNode{3, &ListNode{4, nil}}}}))
}
